### PR TITLE
Add password provision and user name provision capabilities when the JIT provision capabilities is enabled

### PR DESCRIPTION
**Pull request by Megala21**
### Proposed changes in this pull request

- First iteration of https://github.com/wso2/product-is/issues/3082
- With this PR under JIT section, three radion buttons are introduced to decide on whether to get the username, password or to not ask anything. With this consent management is introduced.

### When should this PR be merged
N/A

### Follow up actions

- With this PR db migration is being introduced. Hence that need to be tested. 

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/JustInTimeProvisioningConfig.java [hunk 1]
```diff
@@ -56,6 +59,15 @@ public static JustInTimeProvisioningConfig build(OMElement justInTimeProvisionin
                     justInTimeProvisioningConfig.setProvisioningEnabled(Boolean
                             .parseBoolean(element.getText()));
                 }
+            } else if (IdentityApplicationConstants.IS_PASSWORD_PROVISIONING_ENABLED_ELEMENT.equals(elementName)) {
+                if (element.getText() != null && element.getText().trim().length() > 0) {
```
Comment (malithie) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/JustInTimeProvisioningConfig.java]: Can we use StringUtils.isNotBlank() here

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/JustInTimeProvisioningConfig.java [hunk 2]
```diff
@@ -56,6 +59,15 @@ public static JustInTimeProvisioningConfig build(OMElement justInTimeProvisionin
...
+                    justInTimeProvisioningConfig
+                            .setPasswordProvisioningEnabled(Boolean.parseBoolean(element.getText()));
+                }
+            } else if (IdentityApplicationConstants.ALLOW_MODIFY_USERNAME.equals(elementName)) {
+                if (element.getText() != null && element.getText().trim().length() > 0) {
```
Comment (malithie) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/JustInTimeProvisioningConfig.java]: Can we use StringUtils.isNotBlank() here

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/impl/DefaultProvisioningHandler.java [hunk 3]
```diff
@@ -138,12 +138,23 @@ public void handle(List<String> roles, String subject, Map<String, String> attri
...
                 }
 
-            } else {
+                UserProfileAdmin userProfileAdmin = UserProfileAdmin.getInstance();
 
-                userStoreManager.addUser(username, generatePassword(), addingRoles.toArray(
+                if (StringUtils.isEmpty(userProfileAdmin.getNameAssociatedWith(idp, subjectVal))) {
+                    // Associate User
+                    associateUser(username, userStoreDomain, tenantDomain, subjectVal, idp);
+                }
```
Comment (malithie) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/impl/DefaultProvisioningHandler.java]: Will this work without a tenant flow. AFAIK, UserProfileAdmin identifies tenant from the realm not from the username

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/PostAuthAssociationHandler.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+                        .getAuthenticatedSubjectIdentifier();
+
+                if (stepConfig.isSubjectIdentifierStep()) {
+                    // there can be only step for subject attributes.
+                    String associatedID = null;
+
+                    // now we know the value of the subject - from the external identity provider.
+                    if (sequenceConfig.getApplicationConfig().isAlwaysSendMappedLocalSubjectId()) {
+
+                        // okay - now we need to find out the corresponding mapped local subject
```
Comment (lahirus) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/PostAuthAssociationHandler.java]: Could you re evaluate the comment here? Maybe remove 'ok' and make it a single line.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/PostAuthAssociationHandler.java [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+                        if (log.isDebugEnabled()) {
+                            log.debug("Authenticated User: " + sequenceConfig.getAuthenticatedUser()
+                                    .getAuthenticatedSubjectIdentifier());
+                            log.debug("Authenticated User Tenant Domain: " + tenantDomain);
+                        }
+
+                    } else {
+                        sequenceConfig.setAuthenticatedUser(new AuthenticatedUser(stepConfig.getAuthenticatedUser()));
+
```
Comment (lahirus) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/PostAuthAssociationHandler.java]: Avoid new lines 214, 216, 218, and 220

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/PostAuthenticatedSubjectIdentifierHandler.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+                if (StringUtils.isNotBlank(sequenceConfig.getAuthenticatedUser().getUserName())) {
+                    sequenceConfig.getAuthenticatedUser().setAuthenticatedSubjectIdentifier(
+                            sequenceConfig.getAuthenticatedUser().getUsernameAsSubjectIdentifier(
+                                    sequenceConfig.getApplicationConfig()
+                                            .isUseUserstoreDomainInLocalSubjectIdentifier(),
+                                    sequenceConfig.getApplicationConfig().isUseTenantDomainInLocalSubjectIdentifier()));
+                }
+            }
+
+        } else {
```
Comment (lahirus) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/PostAuthenticatedSubjectIdentifierHandler.java]: Make it else if ()

**Comment by Megala21** [comment]
Since I am allocated for a different task, will close this PR temporarily and will re-open after finishing the current allocation.

**Pull request closed without merging**